### PR TITLE
Adds foundation for the admin API

### PR DESCRIPTION
**Pull request by m1kola**
Depends on:
* https://github.com/Azure/ARO-RP/pull/119
* https://github.com/Azure/ARO-RP/pull/124
* https://github.com/Azure/ARO-RP/pull/145
* https://github.com/Azure/ARO-RP/pull/183

### CODE REVIEW

**Comment by m1kola** [comment]
Needs a rebase on top of https://github.com/Azure/ARO-RP/pull/119

File: pkg/api/admin/openshiftcluster_validatestatic_test.go [hunk 1]
```diff
@@ -0,0 +1,632 @@
+package admin
```
Comment (m1kola) [inline pkg/api/admin/openshiftcluster_validatestatic_test.go]: Copy from `v20191231preview`, but:

* "empty clientSecret invalid" case got removed (no `clientSecret` in admin api)
* Worker count change is not currently allowed
* Tags are not allowed (needs clarification: see the comment in `pkg/api/admin/openshiftcluster.go`)

File: pkg/api/admin/register.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/Azure/ARO-RP/pkg/env"
+)
+
+const (
+	resourceProviderNamespace = "Microsoft.RedHatOpenShift"
+	resourceType              = "openShiftClusters"
+)
+
+func init() {
+	api.APIs["admin"] = &api.Version{
```
Comment (m1kola) [inline pkg/api/admin/register.go]: Copy from `v20191231preview`. The only difference is this line with the api version name

File: pkg/api/admin/openshiftclustercredentials_convert.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import (
+	"github.com/Azure/ARO-RP/pkg/api"
+)
+
+type openShiftClusterCredentialsConverter struct{}
+
+// ToExternal is supposed to return a struct with admin username and password,
+// but we do not expose secrets in the admin API, so we just return an empty struct
+func (*openShiftClusterCredentialsConverter) ToExternal(*api.OpenShiftCluster) interface{} {
+	return struct{}{}
```
Comment (m1kola) [inline pkg/api/admin/openshiftclustercredentials_convert.go]: This means that post to list credentials will receve an empty json object. I don't think that we will have any credentials that we will want to list here in the future. Probably we should look into disabling the list credentials route for the admin API completely?

File: pkg/api/v20191231preview/openshiftcluster.go [hunk 4]
```diff
@@ -68,11 +68,12 @@ type ProvisioningState string
...
 // ProvisioningState constants
 const (
-	ProvisioningStateCreating  ProvisioningState = "Creating"
-	ProvisioningStateUpdating  ProvisioningState = "Updating"
-	ProvisioningStateDeleting  ProvisioningState = "Deleting"
-	ProvisioningStateSucceeded ProvisioningState = "Succeeded"
-	ProvisioningStateFailed    ProvisioningState = "Failed"
+	ProvisioningStateCreating      ProvisioningState = "Creating"
+	ProvisioningStateUpdating      ProvisioningState = "Updating"
+	ProvisioningStateAdminUpdating ProvisioningState = "AdminUpdating"
```
Comment (m1kola) [inline pkg/api/v20191231preview/openshiftcluster.go]: We need it here so user can see in when doing get.

File: pkg/frontend/openshiftcluster_putorpatch.go [hunk 5]
```diff
@@ -85,7 +85,8 @@ func (f *frontend) _putOrPatchOpenShiftCluster(ctx context.Context, r *http.Requ
 		switch doc.OpenShiftCluster.Properties.FailedProvisioningState {
 		case api.ProvisioningStateCreating:
 			return nil, api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeRequestNotAllowed, "", "Request is not allowed on cluster whose creation failed. Delete the cluster.")
-		case api.ProvisioningStateUpdating:
+		case api.ProvisioningStateUpdating, api.ProvisioningStateAdminUpdating:
+			// TODO: Discuss. Do we want users to be able to do operations on clusters in api.ProvisioningStateAdminUpdating failed state
```
Comment (m1kola) [inline pkg/frontend/openshiftcluster_putorpatch.go]: Note the comment

File: pkg/frontend/openshiftcluster_putorpatch.go [hunk 6]
```diff
@@ -139,7 +140,13 @@ func (f *frontend) _putOrPatchOpenShiftCluster(ctx context.Context, r *http.Requ
 		}
 
 	} else {
-		doc.OpenShiftCluster.Properties.ProvisioningState = api.ProvisioningStateUpdating
+		// TODO: Get rid of the special case
```
Comment (m1kola) [inline pkg/frontend/openshiftcluster_putorpatch.go]: I woild like to do it in a separate PR: this one is already big.

File: pkg/api/admin/openshiftcluster.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
+package admin
+
+import (
+	"time"
+)
+
+// Copyright (c) Microsoft Corporation.
```
Comment (jim-minter) [inline pkg/api/admin/openshiftcluster.go]: please move above the import

File: pkg/api/admin/openshiftcluster.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+	OpenShiftClusters []*OpenShiftCluster `json:"value"`
+}
+
+// OpenShiftCluster represents an OpenShift cluster
+type OpenShiftCluster struct {
+	ID         string     `json:"id,omitempty" mutable:"case"`
+	Name       string     `json:"name,omitempty" mutable:"case"`
+	Type       string     `json:"type,omitempty" mutable:"case"`
+	Location   string     `json:"location,omitempty"`
+	Tags       Tags       `json:"tags,omitempty"`
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/api/admin/openshiftcluster.go]: it should not be mutable in the admin api

File: pkg/api/admin/openshiftcluster.go [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+// ConsoleProfile represents a console profile.
+type ConsoleProfile struct {
+	URL string `json:"url,omitempty"`
+}
+
+// ServicePrincipalProfile represents a service principal profile.
+type ServicePrincipalProfile struct {
+	TenantID string `json:"tenantId,omitempty"`
+	ClientID string `json:"clientId,omitempty"`
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/api/admin/openshiftcluster.go]: I think that's fine

File: pkg/api/admin/openshiftcluster.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+	VMSizeStandardD8sV3 VMSize = "Standard_D8s_v3"
+)
+
+// WorkerProfile represents a worker profile
+type WorkerProfile struct {
+	Name       string `json:"name,omitempty"`
+	VMSize     VMSize `json:"vmSize,omitempty"`
+	DiskSizeGB int    `json:"diskSizeGB,omitempty"`
+	SubnetID   string `json:"subnetId,omitempty"`
+	Count      int    `json:"count,omitempty"` // TODO: Add mutable:"true" to allow changes
```
Comment (jim-minter) [inline pkg/api/admin/openshiftcluster.go]: I think the TODO can be removed - this won't be mutable by any API

File: pkg/api/admin/openshiftcluster_example.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
+package admin
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+func exampleOpenShiftCluster() *OpenShiftCluster {
```
Comment (jim-minter) [inline pkg/api/admin/openshiftcluster_example.go]: I keep wondering if we should define this in pkg/api and convert it out to a versioned API in the places that it's needed

File: pkg/api/admin/openshiftcluster_validatedynamic.go [hunk 12]
```diff
@@ -0,0 +1,267 @@
+package admin
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/api/admin/openshiftcluster_validatedynamic.go]: Mmm, finding a way to avoid the copy would be good - this already operates on the internal API, perhaps it can be moved to pkg/api?

File: pkg/api/admin/openshiftclustercredentials_convert.go (hunk 3 above)
Comment (jim-minter) [reply pkg/api/admin/openshiftclustercredentials_convert.go]: Agree, perhaps `OpenShiftClusterCredentialsConverter()` can return nil for this API and we can check for that in openshiftclustercredentials_post.go?

File: pkg/frontend/openshiftcluster_putorpatch.go (hunk 5 above)
Comment (jim-minter) [reply pkg/frontend/openshiftcluster_putorpatch.go]: Need to think about that, I agree.

File: pkg/api/admin/openshiftclustercredentials_convert.go (hunk 3 above)
Comment (m1kola) [reply pkg/api/admin/openshiftclustercredentials_convert.go]: I created a separate PR https://github.com/Azure/ARO-RP/pull/124 for this.

File: pkg/api/admin/openshiftcluster_example.go (hunk 11 above)
Comment (mjudeikis) [reply pkg/api/admin/openshiftcluster_example.go]: +1 on this. And I think this should be on the Document Level and if you need `OpenShiftCluster` you just cherry-pick substructure. But I think this is separate issue/PR

File: pkg/api/admin/openshiftcluster_validatestatic.go [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+	location   string
+	resourceID string
+
+	r azure.Resource
+}
+
+// Validate validates an OpenShift cluster
+func (v *openShiftClusterValidator) Static(_oc interface{}, _current *api.OpenShiftCluster) error {
+	if _current == nil {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeRequestNotAllowed, "", "Admin API does not allow cluster creation.")
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply pkg/api/admin/openshiftcluster_validatestatic.go]: `Static(_oc, _current, isAdmin bool)`? or smth like this?

File: pkg/api/admin/openshiftcluster_example.go (hunk 11 above)
Comment (mjudeikis) [reply pkg/api/admin/openshiftcluster_example.go]: Can we move this to something like `/test/api` as we will need to re-use this all around the place

File: pkg/api/admin/openshiftcluster_validatestatic.go [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+	"github.com/Azure/go-autorest/autorest/azure"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/immutable"
+	"github.com/Azure/ARO-RP/pkg/util/subnet"
+)
+
+var (
+	rxResourceGroupID = regexp.MustCompile(`(?i)^/subscriptions/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/resourceGroups/[-a-z0-9_().]{0,89}[-a-z0-9_()]$`)
```
Comment (mjudeikis) [inline pkg/api/admin/openshiftcluster_validatestatic.go]: These look like material to be placed in `pgk/api` and re-used as they should not change. Or change across the board.

Comment (m1kola) [reply pkg/api/admin/openshiftcluster_validatestatic.go]: @mjudeikis Please take a look at https://github.com/Azure/ARO-RP/pull/145. It changes location of the validators and these variables as well.

Comment (m1kola) [reply pkg/api/admin/openshiftcluster_validatestatic.go]: I'll update this PR once we settle on appraoch for #145

File: pkg/api/admin/openshiftcluster_example.go (hunk 11 above)
Comment (m1kola) [reply pkg/api/admin/openshiftcluster_example.go]: Exampels are used in both: tests and swagger example generation. So it's a bit more complicated than just moving things around.

Anyway - I would prefer to deal with it separately from this PR.

File: pkg/frontend/openshiftcluster_putorpatch.go (hunk 5 above)
Comment (jim-minter) [reply pkg/frontend/openshiftcluster_putorpatch.go]: For now, I suggest that we disallow users from updating a cluster in Failed/AdminUpdating.  My rationale is:
a) that if the update fails, the cluster probably goes into Failed/Updating and we lose track that the admin update caused the first failure
b) we can open this up later if we need to

File: pkg/api/admin/openshiftcluster_validatestatic.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (m1kola) [reply pkg/api/admin/openshiftcluster_validatestatic.go]: Because the validator is a API-version specific - we know that we are in the `admin` or in `v20191231preview` version. So there is no need to pass `isAdmin` as an argument.

**Comment by m1kola** [comment]
@jim-minter @mjudeikis please take another look. There is one thing left (https://github.com/Azure/ARO-RP/pull/118#discussion_r370799265) - I'll try to come up with a neat way to get rid of the edge case in a separate PR, but please feel free to suggest options.

File: pkg/api/admin/openshiftcluster_example.go (hunk 11 above)
Comment (jim-minter) [reply pkg/api/admin/openshiftcluster_example.go]: ok - please open a follow-up

File: pkg/api/admin/openshiftcluster_example.go [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+					SubnetID:   "/subscriptions/subscriptionId/resourceGroups/vnetResourceGroup/providers/Microsoft.Network/virtualNetworks/vnet/subnets/worker",
+					Count:      3,
+				},
+			},
+			APIServerProfile: APIServerProfile{
+				Visibility: VisibilityPublic,
+				URL:        "https://api.cluster.location.aroapp.io:6443/",
+				IP:         "1.2.3.4",
+			},
+			IngressProfiles: []IngressProfile{
```
Comment (jim-minter) [inline pkg/api/admin/openshiftcluster_example.go]: Would we expect to see Install, StorageSuffix populated in this struct as well?

File: pkg/api/admin/openshiftcluster_validatestatic.go [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+	if err := sv.validateAPIServerProfile(path+".apiserverProfile", &p.APIServerProfile); err != nil {
+		return err
+	}
+	if len(p.IngressProfiles) != 1 {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, path+".ingressProfiles", "There should be exactly one ingress profile.")
+	}
+	if err := sv.validateIngressProfile(path+".ingressProfiles['"+p.IngressProfiles[0].Name+"']", &p.IngressProfiles[0]); err != nil {
+		return err
+	}
+
```
Comment (jim-minter) [inline pkg/api/admin/openshiftcluster_validatestatic.go]: strictly, it seems like validation for Install and StorageSuffix is missing here.  At the moment none of this is too applicable because mutation is disabled.  I think we should discuss what strategy we want to take - part of me thinks that if this validation can't be triggered right now, perhaps it shouldn't be here?

File: pkg/api/admin/openshiftcluster_example.go (hunk 15 above)
Comment (m1kola) [reply pkg/api/admin/openshiftcluster_example.go]: @jim-minter I'll add `StorageSuffix`, but not `Install`, I think. This example is in `ProvisioningState == ProvisioningStateSucceeded` state where we don't have `Install`.

Comment (jim-minter) [reply pkg/api/admin/openshiftcluster_example.go]: what to do sort of depends on what our static validation + test strategy is

File: pkg/api/admin/openshiftcluster_validatestatic_test.go (hunk 1 above)
Comment (m1kola) [reply pkg/api/admin/openshiftcluster_validatestatic_test.go]: Because we decided to removed a lot of validations from admin static validator - we now only have delta tests here. I added few more cases to cover admin fields.

File: pkg/api/admin/openshiftcluster_validatestatic_test.go [hunk 17]
```diff
@@ -0,0 +1,477 @@
...
+			modify:  func(oc *OpenShiftCluster) { oc.Properties.WorkerProfiles[0].Count++ },
+			wantErr: "400: PropertyChangeNotAllowed: properties.workerProfiles['worker'].count: Changing property 'properties.workerProfiles['worker'].count' is not allowed.",
+		},
+		{
+			name: "install change is not allowed",
+			oc: func() *OpenShiftCluster {
+				return &OpenShiftCluster{
+					Properties: Properties{
+						Install: &Install{
+							Now:   time.Now(),
```
Comment (m1kola) [inline pkg/api/admin/openshiftcluster_validatestatic_test.go]: Noticed this: https://github.com/Azure/ARO-RP/issues/186

File: pkg/frontend/openshiftcluster_putorpatch.go (hunk 6 above)
Comment (jim-minter) [reply pkg/frontend/openshiftcluster_putorpatch.go]: +1, I don't see much of a neater way anyway

**Comment by jim-minter** [comment]
lgtm.  Don't know if you want to take a look at #187 first then update the "install change is not allowed" test, then we merge?

**Comment by m1kola** [comment]
@jim-minter thank you for fixing it. I LGTM'ed #187 and updated tests in this PR. I expect tests to panic, so it will require rebase once #187 merged.

**Comment by jim-minter** [comment]
pushed with rebase
